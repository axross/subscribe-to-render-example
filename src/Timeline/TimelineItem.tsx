import { ClassAttributes, createElement } from 'react';

interface Props {
  post: any;
}

const TimelineItem = ({ post }: ClassAttributes<HTMLElement> & Props) => {
  console.log(post);

  return <div>{post.body}</div>;
};

export default TimelineItem;
