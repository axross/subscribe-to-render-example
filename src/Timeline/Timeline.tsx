import firebase = require('firebase');
import { ClassAttributes, Component, createElement } from 'react';
import TimelineItem from './TimelineItem';

interface Props {
  firestore: firebase.firestore.Firestore;
}

interface State {
  posts: firebase.firestore.DocumentSnapshot[];
}

class Timeline extends Component<ClassAttributes<HTMLElement> & Props, State> {
  private unsubscribe?: () => void;

  public state: State = {
    posts: [],
  };

  public componentDidMount() {
    this.unsubscribe = this.props.firestore.collection('posts').onSnapshot(querySnapshot => {
      this.setState({ posts: querySnapshot.docs });
    });
  }

  public componentWillUnmount() {
    this.unsubscribe!();
  }

  public render() {
    const items = this.state.posts.map(post => <TimelineItem post={post.data()} key={post.id} />);

    return <div>{items}</div>;
  }
}

export default Timeline;
