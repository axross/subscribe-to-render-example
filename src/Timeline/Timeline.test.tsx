import { shallow } from 'enzyme';
import firebase = require('firebase');
import 'firebase/firestore';
import { createElement } from 'react';
import Timeline from './Timeline';

describe('<Timeline>', () => {
  const firestore = firebase
    .initializeApp({
      apiKey: 'dummy',
      authDomain: 'dummy',
      projectId: 'dummy',
    })
    .firestore();

  test('it matches with what it renders', () => {
    jest.spyOn(firestore, 'collection');

    const rendered = shallow(<Timeline firestore={firestore} />);

    expect(rendered).toMatchSnapshot();
  });

  test('it subscribes collection "posts" of firestore and set own state', () => {
    const posts = [createDummyFirestoreDocumentSnapshot('post1'), createDummyFirestoreDocumentSnapshot('post2'), createDummyFirestoreDocumentSnapshot('post3')];
    const onSnapshot = jest.fn(callback => callback({ docs: posts }));
    const collectionFunction = jest.spyOn(firestore, 'collection').mockImplementation(() => ({ onSnapshot }));

    const rendered = shallow(<Timeline firestore={firestore} />);

    expect(collectionFunction).toHaveBeenCalledWith('posts');
    expect(onSnapshot).toHaveBeenCalled();
    expect(rendered.state().posts).toEqual(posts);
    expect(rendered).toMatchSnapshot();
  });
});

const createDummyFirestoreDocumentSnapshot = (id: string, data?: Record<string, any>) => ({
  id,
  data: () => data ? data : Symbol(id),
});
