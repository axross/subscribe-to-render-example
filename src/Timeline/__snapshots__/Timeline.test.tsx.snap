// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Timeline> it matches with what it renders 1`] = `<div />`;

exports[`<Timeline> it subscribes collection "posts" of firestore and set own state 1`] = `
<div>
  <TimelineItem
    key="post1"
    post={Symbol(post1)}
  />
  <TimelineItem
    key="post2"
    post={Symbol(post2)}
  />
  <TimelineItem
    key="post3"
    post={Symbol(post3)}
  />
</div>
`;
